File: CSharp/Services/Solver.cs
using System.Text.RegularExpressions;

/// <summary>
/// Constraint-satisfaction solver: matches rules against extracted test facts.
///
/// For each rule and each test, tries to find a consistent binding of rule variables
/// ($x, $acc, $result, ...) to actual code variable names that satisfies ALL given
/// clauses (in order) and ALL assert clauses.
///
/// Algorithm: backtracking search over given clauses with ordered facts.
/// Typical complexity: O(N^M) where N = facts per test (5-20) and M = given clauses (1-4).
/// </summary>
static class Solver
{
    /// <summary>
    /// Try to match a rule against a single test's facts.
    /// Returns the result with binding details.
    /// </summary>
    public static RuleResult Match(Rule rule, TestFacts test)
    {
        var result = new RuleResult { Rule = rule };
        var bindings = new Dictionary<string, string>();
        var finalBindings = new Dictionary<string, string>();

        if (SolveGiven(rule.Given, 0, test.Facts, 0, bindings, rule.Assert, test, result, finalBindings))
        {
            result.IsImplemented = true;
            result.MatchedTest = test.TestName;
            result.Bindings = finalBindings;
            result.GivenMatched = rule.Given.Select(DescribeGiven).ToList();
            result.AssertMatched = rule.Assert.Select(DescribeAssert).ToList();
        }
        else
        {
            result.IsImplemented = false;
        }

        return result;
    }

    /// <summary>
    /// Try to match a rule against ALL tests. Returns the best result (first match).
    /// </summary>
    public static RuleResult MatchAgainstAll(Rule rule, List<TestFacts> allTests)
    {
        RuleResult? firstMatch = null;
        var allMatched = new List<string>();

        foreach (var test in allTests)
        {
            var result = Match(rule, test);
            if (result.IsImplemented)
            {
                allMatched.Add(test.TestName);
                firstMatch ??= result;
            }
        }

        if (firstMatch != null)
        {
            firstMatch.AllMatchedTests = allMatched;
            return firstMatch;
        }

        // No test matched — report what's missing
        var failed = new RuleResult
        {
            Rule = rule,
            IsImplemented = false,
            GivenMissing = rule.Given.Select(DescribeGiven).ToList(),
            AssertMissing = rule.Assert.Select(DescribeAssert).ToList()
        };
        return failed;
    }

    // ── Backtracking solver for Given clauses ─────────────

    static bool SolveGiven(
        List<GivenClause> givens, int givenIdx,
        List<Fact> facts, int factStartIdx,
        Dictionary<string, string> bindings,
        List<AssertClause> asserts, TestFacts test,
        RuleResult result, Dictionary<string, string> finalBindings)
    {
        // All given clauses matched — check asserts
        if (givenIdx >= givens.Count)
        {
            if (CheckAsserts(asserts, test, bindings))
            {
                // Propagate successful bindings back to caller
                finalBindings.Clear();
                foreach (var kv in bindings) finalBindings[kv.Key] = kv.Value;
                return true;
            }
            return false;
        }

        var clause = givens[givenIdx];

        // Negated clause: verify NO fact matches, then continue without advancing fact index
        if (clause.Negated)
        {
            for (int fi = 0; fi < facts.Count; fi++)
            {
                var probe = TryMatchGiven(clause, facts[fi], bindings, test);
                if (probe != null)
                    return false; // found a match → negation fails
            }
            // No match found → negation succeeds, continue with same factStartIdx
            return SolveGiven(givens, givenIdx + 1, facts, factStartIdx, bindings, asserts, test, result, finalBindings);
        }

        // Try each fact from factStartIdx onward (preserving order)
        for (int fi = factStartIdx; fi < facts.Count; fi++)
        {
            var newBindings = TryMatchGiven(clause, facts[fi], bindings, test);
            if (newBindings != null)
            {
                if (SolveGiven(givens, givenIdx + 1, facts, fi + 1, newBindings, asserts, test, result, finalBindings))
                    return true;
            }
        }

        return false;
    }

    // ── Given clause matching ─────────────────────────────

    static Dictionary<string, string>? TryMatchGiven(GivenClause clause, Fact fact, Dictionary<string, string> bindings, TestFacts test)
    {
        return clause switch
        {
            CreateClause c => TryMatchCreate(c, fact, bindings),
            CallClause c => TryMatchCall(c, fact, bindings, test),
            AssignClause c => TryMatchAssign(c, fact, bindings),
            MockClause c => TryMatchMock(c, fact, bindings),
            _ => null
        };
    }

    static Dictionary<string, string>? TryMatchCreate(CreateClause clause, Fact fact, Dictionary<string, string> bindings)
    {
        if (fact is not CreatedFact cf) return null;

        // Type must match
        if (!TypeMatches(clause.TypePattern, cf.Type)) return null;

        var b = new Dictionary<string, string>(bindings);

        // Bind the variable
        if (!TryBind(b, clause.BindVar, cf.Variable)) return null;

        // When explicit parens used (e.g. Accessor()), require exact arg count match
        if (clause.ExactArgCount && cf.Arguments.Count != clause.ArgBindings.Count)
            return null;

        // Bind constructor arguments
        for (int i = 0; i < clause.ArgBindings.Count; i++)
        {
            if (i >= cf.Arguments.Count) return null;
            var argBind = clause.ArgBindings[i];
            if (!TryBindOrMatch(b, argBind, cf.Arguments[i])) return null;
        }

        return b;
    }

    static Dictionary<string, string>? TryMatchCall(CallClause clause, Fact fact, Dictionary<string, string> bindings, TestFacts test)
    {
        if (fact is not CalledFact cf) return null;

        // Method name must match
        if (!cf.Method.Equals(clause.MethodName, StringComparison.OrdinalIgnoreCase)) return null;

        var b = new Dictionary<string, string>(bindings);

        // Bind receiver if specified
        if (clause.ReceiverBindVar != null)
        {
            if (cf.Receiver == null) return null;
            if (!TryBind(b, clause.ReceiverBindVar, cf.Receiver)) return null;
        }

        // Check onType if specified
        if (clause.OnTypePattern != null)
        {
            if (cf.ReceiverType == null) return null;
            if (!TypeMatches(clause.OnTypePattern, cf.ReceiverType)) return null;
        }

        // Bind result variable if specified
        if (clause.ResultBindVar != null)
        {
            if (cf.ResultVariable == null) return null;
            if (!TryBind(b, clause.ResultBindVar, cf.ResultVariable)) return null;
        }

        // Bind method arguments
        for (int i = 0; i < clause.ArgBindings.Count; i++)
        {
            if (i >= cf.Arguments.Count) return null;
            if (!TryBindOrMatch(b, clause.ArgBindings[i], cf.Arguments[i])) return null;
        }

        // Bind callback assignment target if specified
        if (clause.CallbackBindVar != null)
        {
            if (cf.CallbackAssignTarget == null) return null;
            if (!TryBind(b, clause.CallbackBindVar, cf.CallbackAssignTarget)) return null;
        }

        // Bind callback throw type if specified
        if (clause.ThrowsBindVar != null)
        {
            if (cf.CallbackThrowType == null) return null;
            if (!TryBindOrMatch(b, clause.ThrowsBindVar, cf.CallbackThrowType)) return null;
        }

        return b;
    }

    static Dictionary<string, string>? TryMatchAssign(AssignClause clause, Fact fact, Dictionary<string, string> bindings)
    {
        if (fact is not AssignedFact af) return null;

        // Cast type must match
        if (af.CastType == null) return null;
        if (!TypeMatches(clause.CastTypePattern, af.CastType)) return null;

        var b = new Dictionary<string, string>(bindings);

        // Bind target
        if (!TryBind(b, clause.TargetBindVar, af.Target)) return null;

        // Bind source
        if (!TryBind(b, clause.SourceBindVar, af.Source)) return null;

        return b;
    }

    static Dictionary<string, string>? TryMatchMock(MockClause clause, Fact fact, Dictionary<string, string> bindings)
    {
        if (fact is not MockSetupFact mf) return null;

        // Method name must match
        if (!mf.Method.Equals(clause.MethodName, StringComparison.OrdinalIgnoreCase)) return null;

        // Return value must match (or "any")
        if (!clause.ReturnValue.Equals("any", StringComparison.OrdinalIgnoreCase))
        {
            if (mf.ReturnValue == null) return null;
            if (!mf.ReturnValue.Equals(clause.ReturnValue, StringComparison.OrdinalIgnoreCase))
                return null;
        }

        var b = new Dictionary<string, string>(bindings);

        if (clause.MockBindVar != null)
        {
            if (!TryBind(b, clause.MockBindVar, mf.MockVariable)) return null;
        }

        return b;
    }

    // ── Assert checking ───────────────────────────────────

    static bool CheckAsserts(List<AssertClause> asserts, TestFacts test, Dictionary<string, string> bindings)
    {
        var assertFacts = test.Facts.OfType<AssertedFact>().ToList();
        var aliases = BuildAliasMap(test);

        foreach (var ac in asserts)
        {
            if (!assertFacts.Any(af => AssertMatches(ac, af, bindings, aliases)))
                return false;
        }

        return true;
    }

    /// <summary>
    /// Build variable alias groups from AssignedFacts.
    /// If publicAccessor = accessor, then they are aliases of each other.
    /// </summary>
    static Dictionary<string, HashSet<string>> BuildAliasMap(TestFacts test)
    {
        var map = new Dictionary<string, HashSet<string>>();

        foreach (var af in test.Facts.OfType<AssignedFact>())
        {
            if (!map.ContainsKey(af.Source)) map[af.Source] = new HashSet<string> { af.Source };
            if (!map.ContainsKey(af.Target)) map[af.Target] = new HashSet<string> { af.Target };
            map[af.Source].Add(af.Target);
            map[af.Target].Add(af.Source);
        }

        // Transitive closure
        bool changed = true;
        while (changed)
        {
            changed = false;
            foreach (var (key, group) in map)
            {
                var toAdd = group.SelectMany(a => map.GetValueOrDefault(a, new HashSet<string>()))
                                 .Where(a => !group.Contains(a)).ToList();
                if (toAdd.Count > 0) { group.UnionWith(toAdd); changed = true; }
            }
        }

        return map;
    }

    static bool AssertMatches(AssertClause clause, AssertedFact fact, Dictionary<string, string> bindings, Dictionary<string, HashSet<string>> aliases)
    {
        // Semantic meaning must match (unless "any")
        if (!clause.SemanticMeaning.Equals("any", StringComparison.OrdinalIgnoreCase))
        {
            if (!fact.SemanticMeaning.Equals(clause.SemanticMeaning, StringComparison.OrdinalIgnoreCase))
                return false;
        }

        // No args to check → semantic match is enough
        if (clause.Args.Count == 0)
            return true;

        // Resolve rule args through bindings
        var resolvedArgs = clause.Args.Select(a => ResolveExpression(a, bindings)).ToList();

        // For 2-arg assertions (AreEqual, AreNotEqual), try both orderings
        if (resolvedArgs.Count == 2 && fact.Arguments.Count >= 2)
        {
            return (ArgMatches(resolvedArgs[0], fact.Arguments[0], aliases) && ArgMatches(resolvedArgs[1], fact.Arguments[1], aliases))
                || (ArgMatches(resolvedArgs[0], fact.Arguments[1], aliases) && ArgMatches(resolvedArgs[1], fact.Arguments[0], aliases));
        }

        // For single-arg assertions, check each resolved arg appears somewhere in fact args
        foreach (var resolved in resolvedArgs)
        {
            if (resolved.Equals("any", StringComparison.OrdinalIgnoreCase))
                continue;
            if (!fact.Arguments.Any(fa => ArgMatches(resolved, fa, aliases)))
                return false;
        }

        return true;
    }

    // ── Expression resolution ─────────────────────────────

    /// <summary>
    /// Resolve a rule expression through bindings.
    /// "$acc" → "accessor", "$acc.Value" → "accessor.Value"
    /// Literals and "any" pass through unchanged.
    /// </summary>
    static string ResolveExpression(string expr, Dictionary<string, string> bindings)
    {
        if (!expr.StartsWith("$")) return expr;

        var dotIdx = expr.IndexOf('.');
        string varPart, rest;
        if (dotIdx >= 0)
        {
            varPart = expr[..dotIdx];
            rest = expr[dotIdx..]; // includes the dot
        }
        else
        {
            varPart = expr;
            rest = "";
        }

        if (bindings.TryGetValue(varPart, out var bound))
            return bound + rest;

        return expr; // unresolved, leave as-is
    }

    /// <summary>
    /// Check if a resolved rule argument matches a fact argument.
    /// Supports alias resolution: "accessor.Value" matches "publicAccessor.Value"
    /// if publicAccessor is an alias of accessor.
    /// </summary>
    static bool ArgMatches(string resolved, string factArg, Dictionary<string, HashSet<string>> aliases)
    {
        if (resolved.Equals("any", StringComparison.OrdinalIgnoreCase))
            return true;

        // Exact match
        if (resolved.Equals(factArg, StringComparison.OrdinalIgnoreCase))
            return true;

        // Normalize whitespace
        var rn = resolved.Replace(" ", "");
        var fn = factArg.Replace(" ", "");
        if (rn.Equals(fn, StringComparison.OrdinalIgnoreCase))
            return true;

        // Contains match for lambda expressions: factArg like "() => accessor.GetValueOrThrow()"
        // should match resolved "accessor.GetValueOrThrow" or "accessor.GetValueOrThrow()"
        if (fn.Contains("=>") && fn.Contains(rn, StringComparison.OrdinalIgnoreCase))
            return true;

        // Alias-based matching: "accessor.Value" vs "publicAccessor.Value"
        var rDot = resolved.IndexOf('.');
        var fDot = factArg.IndexOf('.');
        if (rDot > 0 && fDot > 0)
        {
            var rPrefix = resolved[..rDot];
            var rSuffix = resolved[rDot..];
            var fPrefix = factArg[..fDot];
            var fSuffix = factArg[fDot..];

            if (rSuffix.Equals(fSuffix, StringComparison.OrdinalIgnoreCase))
            {
                if (aliases.TryGetValue(rPrefix, out var group) && group.Contains(fPrefix))
                    return true;
            }
        }

        return false;
    }

    // ── Binding helpers ───────────────────────────────────

    /// <summary>Try to bind a rule variable to a value. Fails if already bound to a different value.</summary>
    static bool TryBind(Dictionary<string, string> bindings, string variable, string value)
    {
        if (variable == "any" || variable == "$any") return true;

        if (bindings.TryGetValue(variable, out var existing))
            return existing == value; // must be consistent

        bindings[variable] = value;
        return true;
    }

    /// <summary>Try to bind or match: if it's a $var, bind it. If it's a literal or "any", match directly.</summary>
    static bool TryBindOrMatch(Dictionary<string, string> bindings, string pattern, string actual)
    {
        if (pattern.Equals("any", StringComparison.OrdinalIgnoreCase)) return true;
        if (pattern.StartsWith("$")) return TryBind(bindings, pattern, actual);
        // Literal match
        return pattern.Equals(actual, StringComparison.OrdinalIgnoreCase);
    }

    // ── Type matching ─────────────────────────────────────

    /// <summary>
    /// Match a type pattern against an actual type name.
    /// "Accessor" matches "Accessor", "Accessor&lt;string&gt;", "Accessor&lt;int&gt;"
    /// "Accessor&lt;any&gt;" matches "Accessor&lt;string&gt;", "Accessor&lt;int&gt;"
    /// "IAccessorSetter&lt;any&gt;" matches "IAccessorSetter&lt;object&gt;"
    /// </summary>
    static bool TypeMatches(string pattern, string actual)
    {
        if (pattern.Equals("any", StringComparison.OrdinalIgnoreCase)) return true;

        // Exact match
        if (pattern.Equals(actual, StringComparison.OrdinalIgnoreCase)) return true;

        // Convert pattern to regex: MyType<any> → ^MyType<.+>$
        var escaped = Regex.Escape(pattern);
        // Replace <any> with <.+> (Regex.Escape does NOT escape < and >)
        var regexStr = escaped.Replace("<any>", "<.+>");
        // Also allow pattern "MyType" to match "MyType<whatever>"
        if (!pattern.Contains('<'))
            regexStr += @"(?:<.+>)?";

        regexStr = "^" + regexStr + "$";

        return Regex.IsMatch(actual, regexStr, RegexOptions.IgnoreCase);
    }

    // ── Description helpers ───────────────────────────────

    static string DescribeGiven(GivenClause g)
    {
        var prefix = g.Negated ? "NOT " : "";
        return prefix + (g switch
        {
            CreateClause c => $"create {c.BindVar}: {c.TypePattern}" + (c.ArgBindings.Count > 0 ? $"({string.Join(", ", c.ArgBindings)})" : ""),
            CallClause c => (c.ResultBindVar != null ? $"{c.ResultBindVar} = " : "") +
                            "call: " + (c.ReceiverBindVar != null ? $"{c.ReceiverBindVar}." : "") +
                            c.MethodName + (c.ArgBindings.Count > 0 ? $"({string.Join(", ", c.ArgBindings)})" : "") +
                            (c.OnTypePattern != null ? $" on {c.OnTypePattern}" : "") +
                            (c.CallbackBindVar != null ? $" -> {c.CallbackBindVar}" : "") +
                            (c.ThrowsBindVar != null ? $" throws {c.ThrowsBindVar}" : ""),
            AssignClause c => $"assign {c.TargetBindVar} = {c.SourceBindVar}: {c.CastTypePattern}",
            MockClause c => $"mock: {c.MethodName} returns {c.ReturnValue}",
            _ => g.ToString() ?? ""
        });
    }

    static string DescribeAssert(AssertClause a) =>
        a.SemanticMeaning + (a.Args.Count > 0 ? $": {string.Join(", ", a.Args)}" : "");
}

File: Python/solver.py
"""
Constraint-satisfaction solver: matches rules against extracted test facts.

For each rule and each test, tries to find a consistent binding of rule variables
($x, $obj, $result, ...) to actual code variable names that satisfies ALL given
clauses (in order) and ALL assert clauses.

Algorithm: backtracking search over given clauses with ordered facts.
"""

import re
from models import (
    Rule, RuleResult, TestFacts, Fact,
    GivenClause, CreateClause, CallClause, AssignClause, MockClause,
    AssertClause,
    CreatedFact, CalledFact, AssignedFact, MockSetupFact, AssertedFact,
)


def match(rule: Rule, test: TestFacts) -> RuleResult:
    result = RuleResult(rule=rule)
    final_bindings: dict[str, str] = {}

    if _solve_given(rule.given, 0, test.facts, 0, {}, rule.asserts, test, final_bindings):
        result.is_implemented = True
        result.matched_test = test.test_name
        result.bindings = dict(final_bindings)
        result.given_matched = [_describe_given(g) for g in rule.given]
        result.assert_matched = [_describe_assert(a) for a in rule.asserts]
    return result


def match_against_all(rule: Rule, all_tests: list[TestFacts]) -> RuleResult:
    first_match: RuleResult | None = None
    all_matched: list[str] = []

    for test in all_tests:
        result = match(rule, test)
        if result.is_implemented:
            all_matched.append(test.test_name)
            if first_match is None:
                first_match = result

    if first_match is not None:
        first_match.all_matched_tests = all_matched
        return first_match

    return RuleResult(
        rule=rule,
        is_implemented=False,
        given_missing=[_describe_given(g) for g in rule.given],
        assert_missing=[_describe_assert(a) for a in rule.asserts],
    )


# ── Backtracking solver ──────────────────────────────

def _solve_given(
    givens: list[GivenClause], given_idx: int,
    facts: list[Fact], fact_start: int,
    bindings: dict[str, str],
    asserts: list[AssertClause], test: TestFacts,
    final_bindings: dict[str, str],
) -> bool:
    if given_idx >= len(givens):
        if _check_asserts(asserts, test, bindings):
            final_bindings.clear()
            final_bindings.update(bindings)
            return True
        return False

    clause = givens[given_idx]

    # Negated clause: verify NO fact matches, then continue without advancing fact index
    if clause.negated:
        for fi in range(len(facts)):
            probe = _try_match_given(clause, facts[fi], bindings, test)
            if probe is not None:
                return False  # found a match → negation fails
        # No match found → negation succeeds, continue with same fact_start
        return _solve_given(givens, given_idx + 1, facts, fact_start, bindings, asserts, test, final_bindings)

    for fi in range(fact_start, len(facts)):
        new_bindings = _try_match_given(clause, facts[fi], bindings, test)
        if new_bindings is not None:
            if _solve_given(givens, given_idx + 1, facts, fi + 1, new_bindings, asserts, test, final_bindings):
                return True

    return False


# ── Given clause matching ─────────────────────────────

def _try_match_given(clause: GivenClause, fact: Fact, bindings: dict[str, str], test: TestFacts) -> dict[str, str] | None:
    if isinstance(clause, CreateClause):
        return _try_match_create(clause, fact, bindings)
    if isinstance(clause, CallClause):
        return _try_match_call(clause, fact, bindings, test)
    if isinstance(clause, AssignClause):
        return _try_match_assign(clause, fact, bindings)
    if isinstance(clause, MockClause):
        return _try_match_mock(clause, fact, bindings)
    return None


def _try_match_create(clause: CreateClause, fact: Fact, bindings: dict[str, str]) -> dict[str, str] | None:
    if not isinstance(fact, CreatedFact):
        return None
    if not _type_matches(clause.type_pattern, fact.type_name):
        return None

    b = dict(bindings)
    if not _try_bind(b, clause.bind_var, fact.variable):
        return None

    if clause.exact_arg_count and len(fact.arguments) != len(clause.arg_bindings):
        return None

    for i, arg_bind in enumerate(clause.arg_bindings):
        if i >= len(fact.arguments):
            return None
        if not _try_bind_or_match(b, arg_bind, fact.arguments[i]):
            return None

    return b


def _try_match_call(clause: CallClause, fact: Fact, bindings: dict[str, str], test: TestFacts) -> dict[str, str] | None:
    if not isinstance(fact, CalledFact):
        return None
    if fact.method.lower() != clause.method_name.lower():
        return None

    b = dict(bindings)

    if clause.receiver_bind_var is not None:
        if fact.receiver is None:
            return None
        if not _try_bind(b, clause.receiver_bind_var, fact.receiver):
            return None

    if clause.on_type_pattern is not None:
        if fact.receiver_type is None:
            return None
        if not _type_matches(clause.on_type_pattern, fact.receiver_type):
            return None

    if clause.result_bind_var is not None:
        if fact.result_variable is None:
            return None
        if not _try_bind(b, clause.result_bind_var, fact.result_variable):
            return None

    for i, arg_bind in enumerate(clause.arg_bindings):
        if i >= len(fact.arguments):
            return None
        if not _try_bind_or_match(b, arg_bind, fact.arguments[i]):
            return None

    if clause.callback_bind_var is not None:
        if fact.callback_assign_target is None:
            return None
        if not _try_bind(b, clause.callback_bind_var, fact.callback_assign_target):
            return None

    if clause.throws_bind_var is not None:
        if fact.callback_throw_type is None:
            return None
        if not _try_bind_or_match(b, clause.throws_bind_var, fact.callback_throw_type):
            return None

    return b


def _try_match_assign(clause: AssignClause, fact: Fact, bindings: dict[str, str]) -> dict[str, str] | None:
    if not isinstance(fact, AssignedFact):
        return None
    if fact.cast_type is None:
        return None
    if not _type_matches(clause.cast_type_pattern, fact.cast_type):
        return None

    b = dict(bindings)
    if not _try_bind(b, clause.target_bind_var, fact.target):
        return None
    if not _try_bind(b, clause.source_bind_var, fact.source):
        return None
    return b


def _try_match_mock(clause: MockClause, fact: Fact, bindings: dict[str, str]) -> dict[str, str] | None:
    if not isinstance(fact, MockSetupFact):
        return None
    if fact.method.lower() != clause.method_name.lower():
        return None

    if not clause.return_value.lower() == "any":
        if fact.return_value is None:
            return None
        if fact.return_value.lower() != clause.return_value.lower():
            return None

    b = dict(bindings)
    if clause.mock_bind_var is not None:
        if not _try_bind(b, clause.mock_bind_var, fact.mock_variable):
            return None
    return b


# ── Assert checking ───────────────────────────────────

def _check_asserts(asserts: list[AssertClause], test: TestFacts, bindings: dict[str, str]) -> bool:
    assert_facts = [f for f in test.facts if isinstance(f, AssertedFact)]
    aliases = _build_alias_map(test)

    for ac in asserts:
        if not any(_assert_matches(ac, af, bindings, aliases) for af in assert_facts):
            return False
    return True


def _build_alias_map(test: TestFacts) -> dict[str, set[str]]:
    alias_map: dict[str, set[str]] = {}
    for fact in test.facts:
        if isinstance(fact, AssignedFact):
            alias_map.setdefault(fact.source, {fact.source}).add(fact.target)
            alias_map.setdefault(fact.target, {fact.target}).add(fact.source)

    # Transitive closure
    changed = True
    while changed:
        changed = False
        for key, group in alias_map.items():
            to_add = set()
            for a in group:
                to_add |= alias_map.get(a, set())
            if not to_add.issubset(group):
                group |= to_add
                changed = True

    return alias_map


def _assert_matches(clause: AssertClause, fact: AssertedFact, bindings: dict[str, str], aliases: dict[str, set[str]]) -> bool:
    # Semantic meaning must match
    if clause.semantic_meaning.lower() != "any":
        if fact.semantic_meaning.lower() != clause.semantic_meaning.lower():
            return False

    if not clause.args:
        return True

    resolved = [_resolve_expression(a, bindings) for a in clause.args]

    # For 2-arg assertions, try both orderings
    if len(resolved) == 2 and len(fact.arguments) >= 2:
        return (
            (_arg_matches(resolved[0], fact.arguments[0], aliases) and _arg_matches(resolved[1], fact.arguments[1], aliases))
            or (_arg_matches(resolved[0], fact.arguments[1], aliases) and _arg_matches(resolved[1], fact.arguments[0], aliases))
        )

    # Single-arg or multi-arg: each resolved arg must appear somewhere
    for r in resolved:
        if r.lower() == "any":
            continue
        if not any(_arg_matches(r, fa, aliases) for fa in fact.arguments):
            return False
    return True


# ── Expression resolution ─────────────────────────────

def _resolve_expression(expr: str, bindings: dict[str, str]) -> str:
    if not expr.startswith("$"):
        return expr

    dot_idx = expr.find(".")
    if dot_idx >= 0:
        var_part = expr[:dot_idx]
        rest = expr[dot_idx:]
    else:
        var_part = expr
        rest = ""

    if var_part in bindings:
        return bindings[var_part] + rest
    return expr


def _arg_matches(resolved: str, fact_arg: str, aliases: dict[str, set[str]]) -> bool:
    if resolved.lower() == "any":
        return True
    if resolved.lower() == fact_arg.lower():
        return True

    # Normalize whitespace
    rn = resolved.replace(" ", "")
    fn = fact_arg.replace(" ", "")
    if rn.lower() == fn.lower():
        return True

    # Lambda contains match
    if "lambda" in fn.lower() and rn.lower() in fn.lower():
        return True

    # Alias-based matching
    r_dot = resolved.find(".")
    f_dot = fact_arg.find(".")
    if r_dot > 0 and f_dot > 0:
        r_prefix = resolved[:r_dot]
        r_suffix = resolved[r_dot:]
        f_prefix = fact_arg[:f_dot]
        f_suffix = fact_arg[f_dot:]
        if r_suffix.lower() == f_suffix.lower():
            if r_prefix in aliases and f_prefix in aliases[r_prefix]:
                return True

    return False


# ── Binding helpers ───────────────────────────────────

def _try_bind(bindings: dict[str, str], variable: str, value: str) -> bool:
    if variable in ("any", "$any"):
        return True
    if variable in bindings:
        return bindings[variable] == value
    bindings[variable] = value
    return True


def _try_bind_or_match(bindings: dict[str, str], pattern: str, actual: str) -> bool:
    if pattern.lower() == "any":
        return True
    if pattern.startswith("$"):
        return _try_bind(bindings, pattern, actual)
    # Literal match
    return pattern.lower() == actual.lower()


# ── Type matching ─────────────────────────────────────

def _type_matches(pattern: str, actual: str) -> bool:
    if pattern.lower() == "any":
        return True
    if pattern.lower() == actual.lower():
        return True

    # Convert pattern to regex
    escaped = re.escape(pattern)
    regex_str = escaped.replace(r"\<any\>", r"<.+>")
    # Also allow pattern "MyType" to match "MyType[whatever]" (Python generics)
    regex_str = regex_str.replace(r"\[any\]", r"\[.+\]")
    if "[" not in pattern and "<" not in pattern:
        regex_str += r"(?:\[.+\])?(?:<.+>)?"
    regex_str = "^" + regex_str + "$"

    return bool(re.match(regex_str, actual, re.IGNORECASE))


# ── Description helpers ───────────────────────────────

def _describe_given(g: GivenClause) -> str:
    prefix = "NOT " if g.negated else ""
    if isinstance(g, CreateClause):
        args_str = f"({', '.join(g.arg_bindings)})" if g.arg_bindings else ""
        return prefix + f"create {g.bind_var}: {g.type_pattern}{args_str}"
    if isinstance(g, CallClause):
        parts = []
        if g.result_bind_var:
            parts.append(f"{g.result_bind_var} = ")
        parts.append("call: ")
        if g.receiver_bind_var:
            parts.append(f"{g.receiver_bind_var}.")
        parts.append(g.method_name)
        if g.arg_bindings:
            parts.append(f"({', '.join(g.arg_bindings)})")
        if g.on_type_pattern:
            parts.append(f" on {g.on_type_pattern}")
        if g.callback_bind_var:
            parts.append(f" -> {g.callback_bind_var}")
        if g.throws_bind_var:
            parts.append(f" throws {g.throws_bind_var}")
        return prefix + "".join(parts)
    if isinstance(g, AssignClause):
        return prefix + f"assign {g.target_bind_var} = {g.source_bind_var}: {g.cast_type_pattern}"
    if isinstance(g, MockClause):
        return prefix + f"mock: {g.method_name} returns {g.return_value}"
    return prefix + str(g)


def _describe_assert(a: AssertClause) -> str:
    if a.args:
        return f"{a.semantic_meaning}: {', '.join(a.args)}"
    return a.semantic_meaning
